### PR TITLE
Add migration fallback, improve crash handling

### PR DESCRIPTION
**Pull request by dshokouhi**
* Add a destructive migration fallback so the app never ends in a state where the database will crash due to a missing migration, I know it may never happen but good to have a fallback just in case
* Replace `cursor.count > 0` with `cursor.moveToFirst()` which is a boolean that will return `true` only IF the cursor can move to the first row
* Place entire migration in try and catch block to catch any errors in the migration
* Keep track if migration was successful in order to move the data after we recreate the table
* Perform clean migration in case of bad data
* Creates high priority notification informing user of the database recreation so they can redo sensors and widgets
* Fires an event to HA to help automate the situation better

Should fix the following sentry errors:

This error occurs when the table exists with more than 1 row but we can't access it, bad data that should be recreated

```
java.lang.IllegalStateException: Couldn't read row 0, col -1 from CursorWindow.  Make sure the Cursor is initialized correctly before accessing data from it.
    at android.database.CursorWindow.nativeGetString(CursorWindow.java)
    at android.database.CursorWindow.getString(CursorWindow.java:438)
    at android.database.AbstractWindowedCursor.getString(AbstractWindowedCursor.java:51)
    at io.homeassistant.companion.android.database.AppDatabase$Companion$MIGRATION_9_10$1.migrate(AppDatabase.kt:184)
    at androidx.room.RoomOpenHelper.onUpgrade(RoomOpenHelper.java:99)
    at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.onUpgrade(FrameworkSQLiteOpenHelper.java:177)
    at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:256)
    at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:163)
    at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getWritableSupportDatabase(FrameworkSQLiteOpenHelper.java:145)
    at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper.getWritableDatabase(FrameworkSQLiteOpenHelper.java:106)
    at androidx.room.RoomDatabase.inTransaction(RoomDatabase.java:476)
    at androidx.room.RoomDatabase.assertNotSuspendingTransaction(RoomDatabase.java:281)
    at io.homeassistant.companion.android.database.sensor.SensorDao_Impl.getFull(SensorDao_Impl.java:445)
    at io.homeassistant.companion.android.sensors.SensorReceiver.updateSensors(SensorReceiver.kt:120)
    at io.homeassistant.companion.android.sensors.SensorReceiver$onReceive$1.invokeSuspend(SensorReceiver.kt:96)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

This error exists because somehow, someway we allowed a blob in our database? Bad data:

```
android.database.sqlite.SQLiteBlobTooBigException: Row too big to fit into CursorWindow requiredPos=0, totalRows=9
    at android.database.sqlite.SQLiteConnection.nativeExecuteForCursorWindow(SQLiteConnection.java)
    at android.database.sqlite.SQLiteConnection.executeForCursorWindow(SQLiteConnection.java:986)
    at android.database.sqlite.SQLiteSession.executeForCursorWindow(SQLiteSession.java:858)
    at android.database.sqlite.SQLiteQuery.fillWindow(SQLiteQuery.java:62)
    at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:153)
    at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:140)
    at io.homeassistant.companion.android.database.AppDatabase$Companion$MIGRATION_9_10$1.migrate(AppDatabase.kt:181)
    at androidx.room.RoomOpenHelper.onUpgrade(RoomOpenHelper.java:99)
    at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.onUpgrade(FrameworkSQLiteOpenHelper.java:177)
    at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:436)
    at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:332)
    at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getWritableSupportDatabase(FrameworkSQLiteOpenHelper.java:145)
    at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper.getWritableDatabase(FrameworkSQLiteOpenHelper.java:106)
    at androidx.room.RoomDatabase.inTransaction(RoomDatabase.java:476)
    at androidx.room.RoomDatabase.assertNotSuspendingTransaction(RoomDatabase.java:281)
    at io.homeassistant.companion.android.database.sensor.SensorDao_Impl.getFull(SensorDao_Impl.java:445)
    at io.homeassistant.companion.android.sensors.SensorReceiver.updateSensors(SensorReceiver.kt:120)
    at io.homeassistant.companion.android.sensors.SensorReceiver$onReceive$1.invokeSuspend(SensorReceiver.kt:96)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

In both of these cases occur during migration 9 to 10 and it is safe to recreate it as the user cannot access the app.

We can consider this a breaking change but without it the app won't function until the user clears data so it should at least allow the app to continue to function.

In the future we can consider storing the data to a file so we can attempt to restore from that.

### CODE REVIEW

**Comment by dshokouhi** [comment]
Looks like it should fix this below error too:

```
java.lang.IllegalStateException: Couldn't read row 0, col -1 from CursorWindow.  Make sure the Cursor is initialized correctly before accessing data from it.
    at android.database.CursorWindow.nativeGetString(CursorWindow.java)
    at android.database.CursorWindow.getString(CursorWindow.java:438)
    at android.database.AbstractWindowedCursor.getString(AbstractWindowedCursor.java:51)
    at io.homeassistant.companion.android.database.AppDatabase$Companion$MIGRATION_9_10$1.migrate(AppDatabase.kt:178)
    at androidx.room.RoomOpenHelper.onUpgrade(RoomOpenHelper.java:99)
    at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.onUpgrade(FrameworkSQLiteOpenHelper.java:177)
    at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:338)
    at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:238)
    at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getWritableSupportDatabase(FrameworkSQLiteOpenHelper.java:145)
    at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper.getWritableDatabase(FrameworkSQLiteOpenHelper.java:106)
    at androidx.room.RoomDatabase.inTransaction(RoomDatabase.java:476)
    at androidx.room.RoomDatabase.assertNotSuspendingTransaction(RoomDatabase.java:281)
    at io.homeassistant.companion.android.database.sensor.SensorDao_Impl.get(SensorDao_Impl.java:302)
    at io.homeassistant.companion.android.sensors.SensorReceiver.onReceive(SensorReceiver.kt:78)
    at android.app.ActivityThread.handleReceiver(ActivityThread.java:3187)
    at android.app.ActivityThread.-wrap17
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1672)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loop(Looper.java:164)
    at android.app.ActivityThread.main(ActivityThread.java:6494)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
java.lang.RuntimeException: Unable to start receiver io.homeassistant.companion.android.sensors.SensorReceiver: java.lang.IllegalStateException: Couldn't read row 0, col -1 from CursorWindow.  Make sure the Cursor is initialized correctly before accessing data from it.
    at android.app.ActivityThread.handleReceiver(ActivityThread.java:3194)
    at android.app.ActivityThread.-wrap17
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1672)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loop(Looper.java:164)
    at android.app.ActivityThread.main(ActivityThread.java:6494)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
```

**Comment by JBassett** [comment]
So my only concern is the fact that the user is never informed that we had to blow away the DB.  This means any sensors they had enabled would need to be re-enabled.  We should probably have a dialog that pops up if we have to fallback to destructive.

**Comment by dshokouhi** [comment]
Ok would a toast message work? We could also consider firing a persistent notification in HA to inform better? Like the admin lol

**Comment by dshokouhi** [comment]
Thinking more on this, firing an event would probably be the most useful here so the admin can be notified and they can automate and handle notifying the user. A pop-up may be odd in this case because they can't really act upon it as the data is gone. Maybe a long toast and taking the user to the sensor screen would make the most sense.

**Comment by JBassett** [comment]
I agree the "manager" of the house will likely be the one to fix but I still think we should let the user know somehow.  If a car has engine trouble the check engine light comes on, it doesn't just phone the repair shop.  I think firing an event that can be automated upon is a great idea but I also think we should create a notification warning the user what happened.

**Comment by dshokouhi** [comment]
Yea that makes sense, the error itself is still severe since we lost whatever data we thought we had, even if it was bad data to begin with.  So I think a high priority notification would be good then, we don't want the user to miss something like this either by being silent.  An event is a good thing to have but if you don't know it existed its not that helpful.  So best to bring immediate attention to the user.  This will be its own notification channel as well.

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -69,6 +69,8 @@
 to your home internet.</string>
   <string name="connected_to_home_assistant">Connected to Home Assistant</string>
   <string name="create_template">Create Template</string>
+  <string name="database_migration_failed">Database migration failed, all sensor data has been reset. Please re-enable your sensors.</string>
```
Comment (JBassett) [inline app/src/main/res/values/strings.xml:72]: We also store widget data in the DB, should warn them about those needing to be recreated as well.

File: app/src/main/java/io/homeassistant/companion/android/database/AppDatabase.kt [hunk 2]
```diff
@@ -178,26 +223,55 @@ abstract class AppDatabase : RoomDatabase() {
...
+                        }
+                        runBlocking {
+                            try {
+                                integrationRepository.fireEvent("mobile_app.migration_failed", mapOf())
+                                Log.d(TAG, "Event sent to Home Assistant")
+                            } catch (e: Exception) {
+                                Log.e(TAG, "Unable to send event to Home Assistant", e)
+                                Toast.makeText(appContext, R.string.database_event_failure, Toast.LENGTH_LONG).show()
+                            }
+                        }
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/database/AppDatabase.kt:265]: By catching the exception, the migration won't throw the exception.  That means the the migration technically won't fail and be recreated.  Also it's duplicate code from above.

**Review by JBassett — CHANGES_REQUESTED** [review]

**Review by JBassett — APPROVED** [review]

**Pull request merged**